unit/util/expr_iterator.cpp

```diff
@@ -156,3 +152,76 @@ TEST_CASE("next_sibling_or_parent, next parent ")
   it.next_sibling_or_parent();
   REQUIRE(it==input[0].depth_end());
 }
+
+/// The mutate_root feature of depth_iteratort can be useful when you have an
+/// `exprt` and want to depth iterate its first operand. As part of that
+/// iteration you may or may not decide to mutate one of the children,
+/// depending on the state of those children. If you do not decide to mutate
+/// a child then you probably don't want to call the non-const version of
+/// `op0()` because that will break sharing, so you create the depth iterator
+/// with the const `exprt` returned from the const invocation of `op0()`, but
+/// if you decide during iteration that you do want to mutate a child then
+/// you can call the non-const version of `op0()` on the original `exprt` in
+/// order to get a non-const `exprt` that the iterator can copy-on-write
+/// update to change the child. At this point the iterator needs to be
+/// informed that it is now safe to write to the `exprt` it contains. This is
+/// achieved by providing a call-back function to the iterator.
+SCENARIO("depth_iterator_mutate_root",
+         "[core][utils][depth_iterator]")
+{
+  GIVEN("A sample expression with a child with id() == ID_1")
+  {
+    // Set up test expression
+    exprt test_expr;
+    exprt test_subexpr;
+    exprt test_subsubexpr(ID_1);
+    test_subexpr.move_to_operands(test_subsubexpr);
+    test_expr.move_to_operands(test_subexpr);
+    WHEN("Iteration occurs without mutation")
+    {
+      // Create shared copies
+      exprt original_expr = test_expr;
+      exprt expr = original_expr;
+      THEN("Shared copy should return object with same address from read()")
+      {
+        REQUIRE(&expr.read() == &original_expr.read());
+      }
+      // Create iterator on first operand
+      const exprt &root = static_cast<const exprt &>(expr).op0();
+      depth_iteratort it = root.depth_begin(
+        [&expr](const exprt &root) -> exprt & { return expr.op0(); });
+      for( ; it != root.depth_cend(); ++it)
+      {
+        if(it->id() == ID_0)
+          it.mutate().id(ID_1);
+      }
+      THEN("No breaking of sharing should have occurred")
+      {
+        REQUIRE(&expr.read() == &original_expr.read());
+      }
+    }
+    WHEN("Iteration occurs with mutation")
+    {
+      // Create shared copies
+      exprt original_expr = test_expr;
+      exprt expr = original_expr;
+      THEN("Shared copy should return object with same address from read()")
+      {
+        REQUIRE(&expr.read() == &original_expr.read());
+      }
+      // Create iterator on first operand
+      const exprt &root = static_cast<const exprt &>(expr).op0();
+      depth_iteratort it = root.depth_begin(
+        [&expr](const exprt &root) -> exprt & { return expr.op0(); });
```

Comment:
If I understand correctly, the const-to-non-const function passed here should return an `irept` whose pointed `dt` equals the `dt` pointed by the `irept` being passed as argument to the function (as per the `INVARIANT` in `mutate`). Now, you are going to trigger the modification (in the `if` within the loop below) on the `test_subsubexpr`, and the expression being returned is the `test_subexpr`. That should violate the invariant.Is this correct? (Travis has not yet passed, so I cannot confirm)

Reply:
Plus, is this the archetypal use of the `mutate_root` function? Shouldn't it always just do an "unsafe" cast of the received `exprt` removing constness?

Reply:
It should never do an unsafe cast. If you don't have a read/write exprt then you can't write to it. You need to get one from somewhere. Perhaps this assumption is a counter-argument to @smowton's idea that this argument should be passed?

Reply:
I've changed the function not to take the const version to avoid confusion and added lots of comment/restructuring to the test to make it clear what's going on.